File: app/src/main/java/com/example/commutingapp/MainScreen.java
package com.example.commutingapp;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import com.facebook.AccessToken;
import com.facebook.AccessTokenTracker;
import com.facebook.login.LoginManager;
import com.google.firebase.auth.UserInfo;

import java.util.ArrayList;
import java.util.List;

import FirebaseUserManager.FirebaseUserManager;
import Logger.CustomToastMessage;
import MenuButtons.BackButtonDoubleClicked;
import MenuButtons.CustomBackButton;
import MenuButtons.backButton;
import Screen.ScreenDimension;

public class MainScreen extends AppCompatActivity implements BackButtonDoubleClicked {

    private CustomToastMessage toastMessageBackButton;
    private TextView nameTextView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        new ScreenDimension(getWindow()).windowToFullScreen();
        setContentView(R.layout.activity_main_screen);
        toastMessageBackButton = new CustomToastMessage(this, getString(R.string.doubleTappedMessage), 10);
        nameTextView = findViewById(R.id.nameTextView);


        FirebaseUserManager.initializeFirebase();
        checkFacebookTokenIfExpired();
       displayName();

    }


    private String getName() {
        for (UserInfo userInfo : FirebaseUserManager.getFirebaseUserInstance().getProviderData()) {
            if (userInfo.equals("facebook.com")) {
                return FirebaseUserManager.getFirebaseUserInstance().getDisplayName();
            }
        }
        return getFilteredEmail(FirebaseUserManager.getFirebaseUserInstance().getEmail());
    }
    private void displayName() {
      nameTextView.setText(getName());
      Log.e("Result from getName ",getName());

    }
    private String getFilteredEmail(String userEmail) {

        List<String> emailLists = getEmailExtensions();
        if (userEmail != null) {
            for (int counter = 0; counter < emailLists.size(); ++counter) {
                String emailExtension = emailLists.get(counter);
                if (userEmail.contains(emailExtension)) {
                    return userEmail.replaceAll(emailExtension, "");
                }
            }
        }

    return userEmail;
    }
    private List<String> getEmailExtensions() {
        List<String> list = new ArrayList<>();
        list.add("@gmail.com");
        list.add("@protonmail.ch");
        list.add("@yahoo.com");
        list.add("@hotmail.com");
        list.add("@outlook.com");
        return list;
    }
    private void checkFacebookTokenIfExpired() {
        new AccessTokenTracker() {
            @Override
            protected void onCurrentAccessTokenChanged(AccessToken oldAccessToken, AccessToken currentAccessToken) {
                if (currentAccessToken == null) {
                    showExpiredTokenDialog();
                    Log.e(getClass().getName(), "Token is Expired");
                }

            }
        };
    }
    public void LogoutButtonClicked(View view) {
        Log.e(getClass().getName(), "Logging out!");
        putUserToLoginFlow();
    }
    private void signOutAccount() {
        LoginManager.getInstance().logOut();
        FirebaseUserManager.getFirebaseAuthInstance().signOut();
    }
    private void putUserToLoginFlow() {
        signOutAccount();
        showSignInForm();
    }
    private void showSignInForm() {
        startActivity(new Intent(this, SignIn.class));
        finish();
    }
    private void showExpiredTokenDialog(){
        startActivity(new Intent(this,TokenExpired.class));
        finish();
    }
    @Override
    public void onBackPressed() {
        backButtonClicked();
    }
    @Override
    public void backButtonClicked() {
        new CustomBackButton(() -> {
            if (backButton.isDoubleTapped()) {
                toastMessageBackButton.hideToast();
                super.onBackPressed();
                return;
            }
            toastMessageBackButton.showToast();
            backButton.registerFirstClick();
        }).backButtonIsClicked();

    }
}

File: app/src/main/java/com/example/commutingapp/MainScreen.kt
package com.example.commutingapp

import FirebaseUserManager.FirebaseManager
import Logger.CustomToastMessage
import MenuButtons.CustomBackButton
import UI.ActivitySwitcher
import UI.AttributesInitializer
import UI.BindingDestroyer
import UI.ScreenDimension
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.example.commutingapp.databinding.ActivityMainScreenBinding
import com.facebook.login.LoginManager
import com.google.firebase.auth.FacebookAuthProvider
import com.google.firebase.auth.GoogleAuthProvider
import com.google.firebase.auth.UserInfo
import java.util.*

class MainScreen : AppCompatActivity(),AttributesInitializer,BindingDestroyer {
    private var toastMessageBackButton: CustomToastMessage? = null
    private var activityMainScreenBinding: ActivityMainScreenBinding? = null
    private lateinit var userInfo: MutableList<out UserInfo>
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initializeAttributes()
        FirebaseManager.initializeFirebaseApp()

    }

    override fun onDestroy() {
        super.onDestroy()
        destroyBinding()
    }
    override fun onStart() {
        super.onStart()
        displayUserProfileName()
    }
    override fun initializeAttributes() {
        ScreenDimension(window).setWindowToFullScreen()
        activityMainScreenBinding = ActivityMainScreenBinding.inflate(layoutInflater)
        setContentView(activityMainScreenBinding?.root)
        toastMessageBackButton = CustomToastMessage(this, getString(R.string.doubleTappedMessage), 10)
        userInfo = FirebaseManager.getFirebaseUserInstance().providerData
    }
    override fun destroyBinding() {
        activityMainScreenBinding = null
    }



    private val userProfileName: String?
        get() {
            for (user in userInfo) {
                if (userSignInViaFacebook(user.providerId) || userSignInViaGoogle(user.providerId)) {
                    Log.e("User id from main screen",user.providerId)
                    return FirebaseManager.getFirebaseUserInstance().displayName
                }
            }
            return filterEmailAddress(FirebaseManager.getFirebaseUserInstance().email)
        }

    private fun userSignInViaFacebook(userProviderId:String) = userProviderId == FacebookAuthProvider.PROVIDER_ID
    private fun userSignInViaGoogle(userProviderId:String) = userProviderId == GoogleAuthProvider.PROVIDER_ID



    private fun displayUserProfileName() {
        activityMainScreenBinding?.nameTextView?.text = userProfileName
    }

    private fun filterEmailAddress(userEmail: String?): String? {
       val emailLists = emailExtensions
       userEmail?.let{
         for (counter in emailLists.indices) {
           val emailExtension = emailLists[counter]
              if (userEmail.contains(emailExtension)) {
                 return userEmail.replace(emailExtension.toRegex(), "")
              }
           }
        }
        return userEmail
    }

    private val emailExtensions: List<String>
        get() {
            val list: MutableList<String> = ArrayList()
            list.add("@gmail.com")
            list.add("@protonmail.ch")
            list.add("@yahoo.com")
            list.add("@hotmail.com")
            list.add("@outlook.com")
            return list
        }

    fun logoutButtonIsClicked(view: View?) {
        Log.e(javaClass.name, "Logging out!")
        putUserToLoginFlow()
    }

    private fun signOutAccount() {
        LoginManager.getInstance().logOut()
        FirebaseManager.getFirebaseAuthInstance().signOut()
    }

    private fun putUserToLoginFlow() {
        signOutAccount()
        showSignInActivity()
    }

    private fun showSignInActivity() {

        ActivitySwitcher.startActivityOf(this,this,SignIn::class.java)
    }

    override fun onBackPressed() {
       CustomBackButton(this,this).applyDoubleClickToExit()
    }



}